src/services/forging.ts

```diff
@@ -1,59 +1,50 @@
 import store from "@/store";
-import { IDelegate } from "../interfaces";
-
-const ForgeStatus: { [key: string]: number } = Object.freeze({
-  FORGING: 0,
-  MISSING: 1,
-  NOT_FORGING: 2,
-  NEVER_FORGED: 3,
-});
+import { IDelegate, IRoundDelegate } from "../interfaces";
+import { ForgingStatus } from "../enums";
+import { roundFromHeight } from "@/utils";
 
 class ForgingService {
-  public status(delegate: IDelegate, height: number): number {
-    let forgingStatus: number;
-
+  public status(delegate: IDelegate, height: number, previousDelegates: IRoundDelegate[]): ForgingStatus {
     if (delegate.blocks.last === undefined) {
-      forgingStatus = ForgeStatus.NEVER_FORGED;
-    } else {
-      const networkRound = this.round(height);
-      const lastBlockRound = this.round(delegate.blocks.last.height);
-      const roundDelta = networkRound - lastBlockRound;
-
-      switch (roundDelta) {
-        case 0:
-        case 1: {
-          forgingStatus = ForgeStatus.FORGING;
-          break;
-        }
-        case 2: {
-          forgingStatus = ForgeStatus.MISSING;
-          break;
-        }
-        default: {
-          forgingStatus = ForgeStatus.NOT_FORGING;
-          break;
-        }
-      }
+      return ForgingStatus.NeverForged;
     }
 
-    return forgingStatus;
-  }
+    const networkRound = roundFromHeight(height);
+    const lastBlockRound = roundFromHeight(delegate.blocks.last.height);
+    const roundDelta = networkRound - lastBlockRound;
 
-  public round(height: number): number {
-    if (isNaN(height)) {
-      return 0;
+    if (!roundDelta) {
+      return ForgingStatus.Forging;
     }
 
-    const activeDelegates: number = store.getters["network/activeDelegates"];
+    if (!previousDelegates.find(el => el.publicKey === delegate.publicKey)) {
+      return ForgingStatus.BecameActive;
+    }
 
-    return Math.floor(height / activeDelegates) + (height % activeDelegates > 0 ? 1 : 0);
+    switch (roundDelta) {
+      // the delegate is either still waiting for their slot or has missed it
+      case 1: {
+        return ForgingStatus.Forging;
+      }
+      // the delegate has missed their slot in the previous round and is now
+      // either still waiting for their slot or has missed their slot
```

Comment:
failed to forge in the previous round, instead of missing slots / blocks as the delegate could miss a slot and forge again in the same round :danstop:

Reply:
will be rephrased